fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,476 @@
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
+        let last_committed_finality = self.last_committed_finality.read()?;
+
+        // parent height is not ready to be proposed yet
+        if height <= last_committed_finality.height {
+            return Ok(None);
+        }
+
+        // prepare block hash and validator set
+        let height_data = self.parent_view_data.height_data.read()?;
+        let block_hash = if let Some(v) = height_data.get_value(height) {
+            let (block_hash, _, _) = v;
+            block_hash.clone()
+        } else {
+            return abort(Error::HeightNotFoundInCache(height));
+        };
+
+        Ok(Some(IPCParentFinality { height, block_hash }))
+    }
+
+    fn check_proposal(&self, proposal: &IPCParentFinality) -> StmResult<(), Error> {
+        self.check_height(proposal)?;
+        self.check_block_hash(proposal)
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) -> StmResult<(), Error> {
+        // the height to clear
+        let height = finality.height;
+
+        self.parent_view_data.height_data.update(|mut cache| {
+            cache.remove_key_below(height + 1);
+            cache
+        })?;
+
+        self.last_committed_finality.write(finality.clone())?;
+
+        Ok(())
+    }
+}
+
+impl InMemoryFinalityProvider {
+    pub fn new(config: Config, committed_finality: IPCParentFinality) -> Self {
+        let height_data = SequentialKeyCache::sequential();
+        Self {
+            config,
+            parent_view_data: ParentViewData {
+                height_data: TVar::new(height_data),
+            },
+            last_committed_finality: TVar::new(committed_finality),
+        }
+    }
+
+    fn check_height(&self, proposal: &IPCParentFinality) -> StmResult<(), Error> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        if latest_height < proposal.height {
+            return abort(Error::ExceedingLatestHeight {
+                proposal: proposal.height,
+                parent: latest_height,
+            });
+        }
+
+        let last_committed_finality = self.last_committed_finality.read()?;
+        if proposal.height <= last_committed_finality.height {
+            return abort(Error::HeightAlreadyCommitted(proposal.height));
+        }
+        Ok(())
```

Comment:
This doesn't seem to be the whole picture. [Here](https://github.com/consensus-shipyard/fendermint/pull/206/files#diff-8edd9923a9eb262950fd919a858ece4449d8a41abf401bd3c1e4ac1d63190454R129) you consider final only something that has certain embedding, yet in this method we accept any height such that `last_committed_finality.height < height <= latest_height`. So it doesn't check that this height is indeed final.

Reply:
I guess this one is about how strict we want the checking to be. According to the design doc, this is the check, as it is more tolerant of the block height difference and relying on the proposals as a final check.

Reply:
@adlrocha can you chime in here. If this is true the design doc seems wrong to me.

Reply:
Once again what I am seeing is that the proposal check _does not finality check whatsoever_.